### PR TITLE
prelimary parallelization

### PR DESCRIPTION
**Pull request by Morgan-W**

### CODE REVIEW

File: alertwildfire_live_recording.py [hunk 1]
```diff
@@ -40,25 +40,7 @@
...
-    #timestamp = time.mktime(datetime.datetime.now().timetuple()) -60*timethreshold
-    #current_target = img_archive.getImgPath("./", "test", timestamp)[-23:-4]
-    #while True:
-    #    for folder in goog_helper.driveListFilesByName(googleServices['drive'], settings.alertwildfire_archive):
-    #        for fileobj in goog_helper.driveListFilesByName(googleServices['drive'], folder['id']):
-    #            if fileobj['name'][-23:-4]< current_target:
-    #                logging.error('deleting file', fileobj['name'])
-    #                goog_helper.deleteItem(googleServices['drive'], file_id)
-    #        
-    return True
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Do you feel deletion should not be part of this file?

File: alertwildfire_live_recording.py [hunk 2]
```diff
@@ -113,14 +95,16 @@ def capture_and_record(googleServices, dbManager, outputDir, camera_name):
...
-def fetchAllCameras(googleServices, camera_names_to_watch):
+def fetchAllCameras(camera_names_to_watch):
     """manages the continual observation of a given set of cameras to watch.
     Args:
-        googleServices: google service service (from getGoogleServices())
         camera_names_to_watch (List): list of camera names that are to be watched by this process
     Returns:
         None
     """
+    googleServices = goog_helper.getGoogleServices(settings, [])
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Does moving this code here fix the permission issue you were having with parallel processing?  If so, please add a comment mentioning this, so someone doesn't move it back out in the future.

File: alertwildfire_live_recording.py [hunk 3]
```diff
@@ -146,28 +130,39 @@ def fetchAllCameras(googleServices, camera_names_to_watch):
...
         listofRotatingCameras = [camera["name"] for camera in listofCameras if (camera["name"][-1]=='2') ]
+    if args.parallelize:
+        parallel = args.parallelize
+    else:
+        parallel = False
 
-    fetchAllCameras(googleServices, listofRotatingCameras)
+    if parallel:#having issues
+        num_cameras_per_process = 5
+        camera_bunchs = [listofRotatingCameras[num_cameras_per_process*num:num_cameras_per_process*num+num_cameras_per_process] for num in range(0, math.ceil(len(listofRotatingCameras)/num_cameras_per_process))]
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Please split this super long line into a multi line for loop so it's more readable.

File: alertwildfire_live_recording.py [hunk 4]
```diff
@@ -146,28 +130,39 @@ def fetchAllCameras(googleServices, camera_names_to_watch):
...
         listofCameras = alertwildfire_API.get_all_camera_info()
         listofRotatingCameras = [camera["name"] for camera in listofCameras if (camera["name"][-1]=='2') ]
+    if args.parallelize:
+        parallel = args.parallelize
+    else:
+        parallel = False
 
-    fetchAllCameras(googleServices, listofRotatingCameras)
+    if parallel:#having issues
+        num_cameras_per_process = 5
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Undocumented fixed magic numbers are usually not great.  Either move these as command line arguments or document how you arrived at these values.

File: alertwildfire_live_recording.py [hunk 5]
```diff
@@ -146,28 +130,39 @@ def fetchAllCameras(googleServices, camera_names_to_watch):
...
     else:
         listofCameras = alertwildfire_API.get_all_camera_info()
         listofRotatingCameras = [camera["name"] for camera in listofCameras if (camera["name"][-1]=='2') ]
+    if args.parallelize:
+        parallel = args.parallelize
+    else:
+        parallel = False
 
-    fetchAllCameras(googleServices, listofRotatingCameras)
+    if parallel:#having issues
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Please elaborate the exact issues vs "#having issues".

File: alertwildfire_live_recording.py [hunk 6]
```diff
@@ -102,12 +104,12 @@ def fetchAllCameras(camera_names_to_watch):
...
     """
+    #please do not remove googleServices definintion from this function
+    # it is needed for the parallel processing authentication
     googleServices = goog_helper.getGoogleServices(settings, [])
-
-
     num_of_watched_cameras = len(camera_names_to_watch)
     dbManager = db_manager.DbManager(sqliteFile=settings.db_file,
-                                    psqlHost=settings.psqlHost, psqlDb=settings.psqlDb,
+                                    psqlHost=settings.psqlHost,              psqlDb=settings.psqlDb,
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: seems like stray whitespace got inserted here by mistake.

File: alertwildfire_live_recording.py [hunk 7]
```diff
@@ -54,8 +55,9 @@ def build_name_with_metadata(image_base_name,metadata):
 
     cameraName_chunk = image_base_name[:-23]
     metadata_chunk = 'p'+str(metadata['position']['pan'])+'_t'+str(metadata['position']['tilt'])+'_z'+str(metadata['position']['zoom'])
-    timeStamp_chunk = '__'+image_base_name[-23:]
-    imgname = cameraName_chunk+metadata_chunk+timeStamp_chunk
+    timeStamp_chunk = image_base_name[-23:-4]+'__'
+    fileTag=image_base_name[-4:]
+    imgname = cameraName_chunk+timeStamp_chunk+metadata_chunk+fileTag
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: This is OK for now, but in future changes, we should switch to img_archive.parseFilename() and repack() for this type of formatting work.

File: alertwildfire_live_recording.py [hunk 8]
```diff
@@ -127,40 +129,98 @@ def fetchAllCameras(camera_names_to_watch):
...
     Returns:
         None
     """
     reqArgs = []
     optArgs = [
+        ["c", "cleaning_threshold", "time in hours to store data"],
         ["o", "cameras_overide", "specific cameras to watch"],
-        ["p", "parallelize", "toggle parallelisation"]
+        ["p", "parallelize", "toggle parallelisation"],
+        ["a", "agents", "number of agents to assign for parallelization"]
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: The 'p' option seems redundant with 'a'.  p = a > 1, right?

File: alertwildfire_live_recording.py [hunk 9]
```diff
@@ -127,40 +129,98 @@ def fetchAllCameras(camera_names_to_watch):
...
+        dbManager = db_manager.DbManager(sqliteFile=settings.db_file,
+                                        psqlHost=settings.psqlHost,                  psqlDb=settings.psqlDb,    
+                                        psqlUser=settings.psqlUser, psqlPasswd=settings.psqlPasswd)
+        temporaryDir = tempfile.TemporaryDirectory()
+        tic = time.time()
+        capture_and_record(googleServices, dbManager, temporaryDir.name, test)
+        toc =time.time()-tic
+        # target estimate of camera refresh time
+        target_refresh_time_per_camera = 12#secs
+        num_cameras_per_process = math.floor(target_refresh_time_per_camera / toc)
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: This type of timing based work splitting should be based on average timing over multiple requests, and also periodically repeating the measurement and adapting.  That complexity is not worth it for version 1, but something to keep in mind.

File: alertwildfire_live_recording.py [hunk 10]
```diff
@@ -127,40 +129,98 @@ def fetchAllCameras(camera_names_to_watch):
...
+        
+
+        #divy the cameras
+        camera_bunchs= []
+        num_of_processes_needed  =  math.ceil(len(listofRotatingCameras)/num_cameras_per_process)
+        if not agents:
+            agents = multiprocessing.cpu_count() -2#as not to inhibit computer functionality
+        if num_of_processes_needed>agents:
+            logging.warning('unable to process all cameras on this machine and maintain a target refresh rate of %s seconds, please reduce number of cameras to less than %s', target_refresh_time_per_camera,num_cameras_per_process*agents)
+            return
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: OK, I see what you're doing now.  Personally, I would simplify everything and just require the agents parameter and base things off of that, and collect some stats on how fast we can process based on different agents values.

File: alertwildfire_live_recording.py (hunk 7 above)
Comment (Morgan-W) [reply alertwildfire_live_recording.py]: need to update parseFilename() and repack to handle new format of name

File: alertwildfire_live_recording.py [hunk 11]
```diff
@@ -143,31 +138,90 @@ def fetchAllCameras(googleServices, camera_names_to_watch):
...
+    while True:
+        current_time = time.time()
+        target_time = current_time - timethreshold*60*60
+        SQLcommand="select FileID from archive where timestamp<"+str(target_time)
+        matches = dbManager.query(SQLcommand)
+        for match in matches:
+            goog_helper.deleteBucketObject(googleServices['storage'], settings.archive_storage_bucket, match['fileID'])
+            #does not yet change database
+    
+    return True
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: No point returning True if you're going to keep the forever while loop above.

File: alertwildfire_live_recording.py [hunk 12]
```diff
@@ -143,31 +138,90 @@ def fetchAllCameras(googleServices, camera_names_to_watch):
...
+def cleanup_archive(googleServices, dbManager, timethreshold):
+    """initializes a continual cleaning function that only acts to remove archived data past a given threshold.
+    Args:
+        googleServices: Drive service (from getGoogleServices())
+        dbManager: Database service
+        timethreshold (flt): hours to keep data in archive
+    Returns:
+        continual should never return must be manually killed
+    """
+    while True:
```
Comment (kinshuk) [inline alertwildfire_live_recording.py]: Why make this a forever loop vs. one-time run?  If you are going to keep it as a forever loop, you should at least add a sleep in there (maybe 10 minute?)

**Pull request merged**